### ISSUE TITLE
Clarification regarding preferredTypes, requiredTypes and conditions.

### ISSUE THREAD

**Issue by MrNickArcher** [MicrosoftDocs/powerbi-docs#4239]
May we please have some additional clarification in this documentation regarding preferredTypes and how it interacts with requiredTypes. Should I use one or the other or both? What effect do they have on the UI? Also do they interact with "conditions" in any way? More info about "conditions" would be great please

## Context

Adding "prefferedTypes", "requiredTypes" and "conditions" breaks my table data view mapping (excerpt of capabilities.json shown below). Although it appears valid and there are no error messages, the visual becomes unresponsive and the update function isn't even called (maybe an exception happens outside of my visual's code?). Removing all "conditions", "requiredTypes" and "preferredTypes" causes it to start working again. Some combinations work, but cause the road_number column to map as all null, even though i have carefully confirmed that my test data is set up to meet the requiredTypes.

I have read the existing documentation over and over and I am struggling to understand what I am doing wrong :/
(I am working in "apiVersion": "5.1.0")

```json
{
"dataRoles": [
        {
            "displayName":"Road",
            "name":"road_number",
            "kind":"Grouping",
            "requiredTypes" : [{"text": true}],
            "preferredTypes": [{"text": true}]
        },
        {
            "displayName":"SLK From (km)",
            "name":"slk_from",
            "kind":"Grouping",
            "requiredTypes" : [{"numeric": true}],
            "preferredTypes": [{"numeric": true}]
        },
        {
            "displayName":"SLK To (km)",
            "name":"slk_to",
            "kind":"Grouping",
            "requiredTypes" : [{"numeric": true}],
            "preferredTypes": [{"numeric": true}]
        },
        {
            "displayName":"Carriageway",
            "name":"cwy",
            "kind":"Grouping",
            "requiredTypes" : [{"text": true}],
            "preferredTypes": [{"text": true}]
        },
        {
            "displayName":"Offset",
            "name":"offset",
            "kind":"Grouping",
            "requiredTypes" : [{"numeric": true}],
            "preferredTypes": [{"numeric": true}]
        }
        ,
        {
            "displayName":"Colour",
            "name":"colour",
            "kind":"Grouping",
            "requiredTypes" : [{"text": true}],
            "preferredTypes": [{"text": true}]
        }
],
"dataViewMappings": [
        {
           "conditions": [{
                    "road_number":{"max": 1, "min":1},
                    "slk_from":{"max": 1, "min":1},
                    "slk_to":{"max": 1, "min":1}
                }],
            "table": {
                "rows": {
                    "select": [
                        {"bind": {"to": "road_number"}},
                        {"bind": {"to": "cwy"}},
                        {"bind": {"to": "slk_from"}},
                        {"bind": {"to": "slk_to"}},
                        {"bind": {"to": "offset"}},
                        {"bind": {"to": "colour"}}
                    ]
                }
            }
        }
    ]
}
```

Some combinations result in an invalid data mapping where the road_number is null, even though it is definitely a properly set up as a text type column in my test data.

```text
{
  "rows": [
                [
                    0, //<--slk_from is numeric and works fine
                    2, //<-- slk_to is also fine
                    null  //<-- I expect road_number text here like "H015"
                ],
                [
                    3,
                    5,
                    null
                ],
                [
                    4,
                    7,
                    null
                ],
                [
                    4,
                    8,
                    null
                ]
            ]
}
```

Many thanks

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 1c9533a4-e32f-6641-2530-2ea21b88f444
* Version Independent ID: 7e80ce3b-c2ea-5e60-ef8f-0ec725dba79b
* Content: [Capabilities and properties of Power BI visuals - Power BI](https://learn.microsoft.com/en-us/power-bi/developer/visuals/capabilities)
* Content Source: [powerbi-docs/developer/visuals/capabilities.md](https://github.com/MicrosoftDocs/powerbi-docs/blob/main/powerbi-docs/developer/visuals/capabilities.md)
* Service: **powerbi**
* Sub-service: **powerbi-custom-visuals**
* GitHub Login: @mberdugo
* Microsoft Alias: **monaberdugo**

**Comment by bishalgoswami** [comment]
Thanks for your comment, @MrNickArcher. This issue has been assigned to the doc author for review.

**Comment by mberdugo** [comment]
Thanks for your question, @MrNickArcher. I'll forward this to the engineers and see if they can provide some guidance.

**Comment by mberdugo** [comment]
Hi @MrNickArcher , I'm sorry about the delayed response. To answer some of your questions, 
You can use requiredTypes and/or preferredTypes, or neither. They are both optional. If you check out some of the examples, you see that you can list several required types, meaning that it must be one of the listed types. Then you can define one or more of the required types as the preferredType. 
For more information on conditions, check out the more detailed page on [datamappings](https://learn.microsoft.com/en-us/power-bi/developer/visuals/dataview-mappings#conditions).

As for your specific question, it seems that you defined 3 min/max values in one step so it's imposible to fulfill all 3 at once. Try separating them into different steps so they are fulfilled one at a time as follows:

"conditions": [
                { "one": { "max": 1, "min": 1 }, "two": { "max": 1, "min": 0 },  "three": { "max": 1, "min": 0 } },
                { "one": { "max": 1, "min": 1 }, "two": { "max": 1, "min": 1 },  "three": { "max": 1, "min": 0 } },
                { "one": { "max": 1, "min": 1 }, "two": { "max": 1, "min": 1 },  "three": { "max": 1, "min": 1 } }
            ],

Good luck.

**Comment by MrNickArcher** [comment]
Hi @mberdugo, Thankyou for your time and response :)

> As for your specific question, it seems that you defined 3 min/max values in one step so it's imposible to fulfill all 3 at once. 

I think I understand what you are saying with your example code...; at least one of the `"conditions"` must be satisfy-able during the process of the user drag dropping columns into the field wells?

This seems counter intuitive to me; I only want the visual work when all three field wells are populated with exactly one column. To me it feels like this should work;

```text
"conditions": [
{ "one": { "max": 1, "min": 1 }, "two": { "max": 1, "min": 1 }, "three": { "max": 1, "min": 1 } }
]
```

I will keep trying different things anyway. After I asked this question I did further tinkering and I noticed that a changing some roles from `Grouping` to `GroupingOrMeasure` seemed to affect the behaviour of the "conditions" as well. Maybe it doesn't like having them all as Grouping, I don't want to use measures though.

I have a feeling I might be even be experiencing some versioning issue and reinstalling pbiviz might help. I don't recall having this problem when I built a similar visual serval versions ago.
